### ISSUE TITLE
Suggest/Autocomplete broken with filters?

### ISSUE THREAD

**Issue by tbd-rsm** [TYPO3-Solr/ext-solr#2212]
Hi folks,

I'm using Typo 9.5.3 and Ext:solr on current dev-master (8e4f2fc).

Is it just me or is the autosuggest example broken when filters are configured?
The configured filter seems to get added twice, which Solarium doesn't like.

I have a filter configured as:

```
plugin.tx_solr.search.query.filter {
  noFilerefs = -(subtype_stringS:filerefs)
}
```

and get the following exception for the suggest ajax request:

```
(1/1) Solarium\Exception\InvalidArgumentException
A filterquery must have a unique key value within a query
```

The first time the filter is added by the search form's [data-suggest]-Url.
The second time the filter is added when the search is performed by the suggest controller.

**Form and Request**
The SearchController populates the `additionalFilters` template variable with data from `ApacheSolrForTypo3\Solr\Domain\Search\Query\QueryBuilder::getAdditionalFilters()`:

```php
//ApacheSolrForTypo3\Solr\Controller\SearchController::resultsAction()

$values = [
  'additionalFilters' => $this->searchService->getAdditionalFilters(),
  ...
];
```

```php
//ApacheSolrForTypo3\Solr\Domain\Search\ResultSet\SearchResultSetService::getAdditionalFilters()
return $this->queryBuilder->getAdditionalFilters();
```

That data is added to the form by it's partial.

```xml
<!-- Ext:solr/Resources/Private/Partials/Search/Form.html -->
<s:searchForm id="tx-solr-search-form-pi-results" 
              additionalFilters="{additionalFilters}"
              suggestHeader="{s:translate(key:'suggest_header',default:'Top Results')}">
```

The resulting xhr request url looks like that:

```
https://foo.bar/suche?tx_solr%5BadditionalFilters%5D%5BnoFilerefs%5D=-%28subtype_stringS%3Afilerefs%29&type=7384&tx_solr[callback]=jQuery33107562062830768079_1547047966678&tx_solr%5BqueryString%5D=tes&_=1547047966682
```

**SuggestController**
The suggest controller now performs a search to provide suggestions.
The query for that is built by the QueryBuilder:

```php
//ApacheSolrForTypo3\Solr\Domain\Search\Query\QueryBuilder::buildSuggestQuery()

$this->newSuggestQuery($queryString)
  ->useFiltersFromTypoScript()  // <- this adds the filter from typoscript configuration
  ->useSiteHashFromTypoScript($requestedPageId)
  ->useUserAccessGroups(explode(',', $groupList))
  ->useOmitHeader();

if (!empty($additionalFilters)) {
  $this->useFilterArray($additionalFilters); // <- this adds the same filter from the request url, which leads to the exception
}

return $this->queryToBuild;
```

The QueryBuilder adds the filter twice - once from the config and second from the additionalFilters passed by the request url.
Both times the filters get build by `ApacheSolrForTypo3\Solr\Domain\Search\Query\QueryBuilder::getAdditionalFilters()`.

**Question**
Is this a bug in the example or in the Extension?

I could just remove the additionalFilters from the form. But is this safe in regard to support of other features. I fear not, (for example) at least for other extensions that add filters via Hook/Slot.

Should filters that get added twice with the same key and query possibly get ignored?

**Comment by timohund** [comment]
I

**Comment by tbd-rsm** [comment]
Hi @timohund Sure, I'd be happy to do so.
Thank you!

**Comment by timohund** [comment]
@tbd-rsm I've added the to patches as pr. It would be nice when you could apply them and check if this solves the problem in your installation as well.

**Comment by tbd-rsm** [comment]
Hi @timohund **Looks good!**

Everything works the way I think the patches intend it to do.
Details below.

The patches got testet on 039c82a.
Details below.

### Pre-Check: Error 1 and it's temp fix are still reproducable without patches.

```xml
<!-- default additional filters -->
<s:searchForm id="tx-solr-search-form-pi-results{idPostfix}" additionalFilters="{additionalFilters}">
```
- [x] does not work (as expected)

```xml
<!-- temp fix: reset default additional filters -->
<f:variable name="additionalFilters" value="{null}" />
<s:searchForm id="tx-solr-search-form-pi-results{idPostfix}" additionalFilters="{additionalFilters}">
```
- [x] does work (as expected)

### Test: Error 1 is fixed by #2225 
applied https://patch-diff.githubusercontent.com/raw/TYPO3-Solr/ext-solr/pull/2225.patch

```xml
<!-- default additional filters (now empty) -->
<s:searchForm id="tx-solr-search-form-pi-results{idPostfix}" additionalFilters="{additionalFilters}">
```
- [x] does work (as intended by #2225)
- [x] Suggestions still get filtered correctly.

### Pre-Check Error 2 is still reproducable without #2226 applied

```xml
<!-- additional filters from a custom SearchController -->
<s:searchForm id="tx-solr-search-form-pi-results{idPostfix}" additionalFilters="{additionalFilters}">
```

```php
// custom Search Controller
protected function getAdditionalFilters()
{
  return [
    // The same filter that is configured in TypoScript.
    'noFilerefs' => '-(subtype_stringS:filerefs)',
    // Control Test: A filter that is not configured in TypoScript.
    'noNews' => '-(subtype_stringS:news)',
  ];
}
```
- [x] does not work (as expected)

### Test: Error 2 is fixed by #2226 
applied https://patch-diff.githubusercontent.com/raw/TYPO3-Solr/ext-solr/pull/2226.patch

```xml
<!-- additional filters from a custom SearchController -->
<s:searchForm id="tx-solr-search-form-pi-results{idPostfix}" additionalFilters="{additionalFilters}">
```

```php
// custom Search Controller
protected function getAdditionalFilters()
{
  return [
    // The same filter that is configured in TypoScript.
    'noFilerefs' => '-(subtype_stringS:filerefs)',
    // Control Test: A filter that is not configured in TypoScript.
    'noNews' => '-(subtype_stringS:news)',
  ];
}
```
- [x] does work (as expected)
- [x] Additional test filter is applied too.

Thanks for fixing. Looking forward to see it in master.

**Comment by timohund** [comment]
Resolved with #2225 and #2226
